File: liveDemo/Acc.cpp
#include <iostream>

class Accumulator {
public:
	double accumulated;
	int nElems;

	Accumulator();
	virtual void accept(double x); 	//NEED TO BE VIRTUAL
	virtual double getMean(); 		//NEED TO BE VIRTUAL
};

Accumulator::Accumulator() {}

void Accumulator::accept(double x) {} //dummy implementation
double Accumulator::getMean() {return -1;} //dummy implementation

//------------------------------------------------------

double process(Accumulator* a) {
	for(int i=1;i<5;i++) {
		a->accept(i);
	}
	std::cout << "[C++] number of elements: " << a->nElems << std::endl;
	return a->getMean();
}

File: liveDemo/Acc.js
arithm = {accumulated: 0, nElems: 0}
arithm.accept = function (x) {
	this.accumulated+=x;
	this.nElems++;
}
arithm.getMean = function () {
	return this.accumulated/this.nElems;
}

geom = {accumulated: 1, nElems: 0}
geom.accept = function (x) {
	this.accumulated*=x;
	this.nElems++;
}
geom.getMean = function () {
	return Math.pow(this.accumulated, 1/this.nElems);
}

//--------------------------------

acc_cpp = Polyglot.evalFile("llvm", "Acc.so");

am = acc_cpp.process(arithm)
print("arithmetic mean:", am)
gm = acc_cpp.process(geom)
print("geometric mean:", gm)
